js/foam/dao/AuthenticatedDAO.js

```diff
@@ -0,0 +1,225 @@
+/**
+ * @license
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+CLASS({
+  package: 'foam.dao',
+  name: 'AuthenticatedDAO',
+  // Deliberately does NOT extend AbstractDAO, ProxyDAO, etc. It would be unsafe
+  // for a new method to be added, but not authenticated.
+  properties: [
+    {
+      name: 'model',
+      required: true
+    },
+    {
+      name: 'delegate',
+      required: true
+    },
+    {
+      name: 'authenticator',
+      documentation: 'An instance of $$DOC{ref:"foam.dao.Authenticator"} ' +
+          'that describes how to authenticate each DAO method.',
+      required: true
+    },
+  ],
+
+  methods: [
+    function adelegateFind_(id) {
+      return function(ret) {
+        this.delegate.find(id, {
+          put: ret,
+          error: function() { ret(null); }
+        });
+      }.bind(this);
+    },
+    function put(obj, sink, opt_X) {
+      // On put, ask the authenticator to decorate our object.
+      // It might return null, indicating that the object is owned by another
+      // user, and this put should be denied.
+      var self = this;
+      var principal = opt_X && opt_X.principal;
+      if (!principal) {
+        sink && sink.error && sink.error('Illegal put.');
+        return;
+      }
+
+      aseq(
+        // Always return null if ID is not set; this is a new object.
+        obj.id ? this.adelegateFind_(obj.id) : aconstant(null),
+        function(ret, old) {
+          self.authenticator.massageForPut(ret, principal, old, obj);
+        },
+        function(ret, massaged) {
+          if (massaged === null) {
+            sink && sink.error && sink.error('Illegal put.');
+          } else {
+            self.delegate.put(massaged, {
+              error: function(e) {
+                // Generic error to avoid data leaks.
+                sink && sink.error && sink.error('Internal error in put');
+                ret();
+              },
+              put: function(postPut) {
+                self.authenticator.massageForRead(function(massaged) {
+                  // Even if the massaged value is false, the put didn't fail.
+                  // So we always call sink.put, maybe with null.
+                  sink && sink.put && sink.put(massaged || null);
```

Comment:
I don't think this is right either.  We should never call back to sink.put with null.  No sinks expect this, no DAOs implement this behaviour.  It is reasonable to expect if a DAO calls your sink.put it will pass you an object every time.

I would do massaged || obj, at the very least the user gets the object they passed in originally.